### PR TITLE
Add AutomationEnvironment to the User-Agent

### PR DESCRIPTION
**Pull request by slewis74**
Update to OctopusClient so it includes the BuildEnvironment in the User-Agent header when it recognises it's being used by a TeamCity, Bamboo or Azure DevOps build agent.

The header is now formatted as follows:

`OctopusClient-dotnet/version (system-information) AutomationEnvironment[/version][ octo[ plugin/version]]`

where:
system-information conveys info about the OS environment
AutomationEnvironment is something like NoneOrUnknown, TeamCity, Bamboo, AzureDevOps, or Octopus
`octo` is added if OctopusClient is being used from the `octo.exe` command line.
'plugin' is added if octo is being called from one of our build server extensions.

### CODE REVIEW

File: source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersAzureDevOpsTests.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using Octopus.Client.Model;
+
+namespace Octopus.Client.Tests.Integration.OctopusClient
+{
+    public class BuildServerHeadersAzureDevOpsTests : BuildServerHeadersBaseTests
+    {
+        protected override string EnvironmentVariableName => OctopusCustomHeaders.EnvVar_AzureDevOps;
```
Comment (matt-richardson) [inline source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersAzureDevOpsTests.cs]: Potentially coupled too closely to the implementation here - ie, the value could change and the test would pass, but it'd break when parsing the telemetry.

File: source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersBaseTests.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            });
+        }
+
+        protected abstract string EnvironmentVariableName { get; }
+        protected abstract string EnvironmentVariableValue { get; }
+        internal abstract BuildEnvironment ExpectedBuildEnvironment { get; }
+
+        protected override void SetupEnvironmentVariables()
+        {
+            OctopusCustomHeaders.GetEnvironmentVariable = variableName => variableName == EnvironmentVariableName ? EnvironmentVariableValue : null;
```
Comment (matt-richardson) [inline source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersBaseTests.cs]: this is a little bit of a brain bender

File: source/Octopus.Client.Tests/Integration/OctopusClient/UserAgentTests.cs [hunk 3]
```diff
@@ -22,11 +23,22 @@ public UserAgentTests()
...
+            OctopusCustomHeaders.GetEnvironmentVariable = Environment.GetEnvironmentVariable;
+        }
+
+
         [Test]
         public async Task AsyncClient_ShouldProvideUserAgent_WithNameAndVersion()
         {
             var response = await AsyncClient.Get<TestDto>(TestRootPath);
-            response.UserAgentValue.Should().Be($"{ApiConstants.OctopusUserAgentProductName}/{GetType().GetSemanticVersion().ToNormalizedString()}", "We should set the standard User-Agent header");
+            response.UserAgentValue.Should().Be($"{ApiConstants.OctopusUserAgentProductName}/{GetType().GetSemanticVersion().ToNormalizedString()} {BuildEnvironment.Unspecified}", "We should set the standard User-Agent header");
```
Comment (matt-richardson) [inline source/Octopus.Client.Tests/Integration/OctopusClient/UserAgentTests.cs]: Again, I'd posit that this is too tightly coupled...

File: source/Octopus.Client/Model/BuildEnvironment.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Octopus.Client.Model
+{
+    internal enum BuildEnvironment
+    {
+        Unspecified,
```
Comment (matt-richardson) [inline source/Octopus.Client/Model/BuildEnvironment.cs]: Worth spending half an hour to hack in other options?
Maybe use https://github.com/pvdlg/env-ci as inspiration?

File: source/Octopus.Client/OctopusAsyncClient.cs [hunk 5]
```diff
@@ -126,9 +125,27 @@ public static async Task<IOctopusAsyncClient> Create(OctopusServerEndpoint serve
 #endif
         }
 
-        private static async Task<IOctopusAsyncClient> Create(OctopusServerEndpoint serverEndpoint, OctopusClientOptions options, bool addHandler)
+        internal static async Task<IOctopusAsyncClient> Create(OctopusServerEndpoint serverEndpoint, OctopusClientOptions options, string requestingTool)
         {
-            var client = new OctopusAsyncClient(serverEndpoint, options ?? new OctopusClientOptions(), addHandler);
+#if HTTP_CLIENT_SUPPORTS_SSL_OPTIONS
```
Comment (matt-richardson) [inline source/Octopus.Client/OctopusAsyncClient.cs]: This looks unrelated to the PR?

File: source/Octopus.Client/OctopusClientFactory.cs [hunk 6]
```diff
@@ -8,6 +8,13 @@ namespace Octopus.Client
     /// </summary>
     public class OctopusClientFactory : IOctopusClientFactory
     {
+        private static string requestingTool;
```
Comment (matt-richardson) [inline source/Octopus.Client/OctopusClientFactory.cs]: Not enthused by static methods like this. It'd be better to pass it in in `CliProgram` when configuring the builder.

File: source/Octopus.Client/Model/OctopusCustomHeaders.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using Octopus.Client.Extensions;
+
+namespace Octopus.Client.Model
+{
+    internal class OctopusCustomHeaders
+    {
+        internal static readonly string EnvVar_TeamCity = "TEAMCITY_VERSION";
```
Comment (matt-richardson) [inline source/Octopus.Client/Model/OctopusCustomHeaders.cs]: Consider a hashtable / dictionary here, so you can just look through that in `DetermineBuildEvironment` instead of an if for every case

File: source/Octopus.Client/Model/BuildEnvironment.cs (hunk 4 above)
Comment (matt-richardson) [reply source/Octopus.Client/Model/BuildEnvironment.cs]: I'd definitely add in Jenkins, appveyor, bitbucket, travis at least. The more the merrier...

File: source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersBaseTests.cs (hunk 2 above)
Comment (slewis74) [reply source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersBaseTests.cs]: Yes. Yes it is. I had to do this, because it relies on Environment variables that our build servers will always set to be the ones for TeamCity, because that's what they are :) So the live code uses the real `Environment.GetEnvironmentVariable` as that function delegate and the tests can override to mock the environment variables

File: source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersAzureDevOpsTests.cs (hunk 1 above)
Comment (slewis74) [reply source/Octopus.Client.Tests/Integration/OctopusClient/BuildServerHeadersAzureDevOpsTests.cs]: Yep, good point, I'll change that

File: source/Octopus.Client/Model/BuildEnvironment.cs (hunk 4 above)
Comment (slewis74) [reply source/Octopus.Client/Model/BuildEnvironment.cs]: yep, I'll add the extras.

File: source/Octopus.Client/OctopusAsyncClient.cs (hunk 5 above)
Comment (slewis74) [reply source/Octopus.Client/OctopusAsyncClient.cs]: it is related, I added a new private method above the original and also added a new defaulted parameter to the original. Git is visualising that in a confusing manner

File: source/Octopus.Client/Model/OctopusCustomHeaders.cs (hunk 7 above)
Comment (slewis74) [reply source/Octopus.Client/Model/OctopusCustomHeaders.cs]: 👍
